subprojects/core/src/main/java/org/gradle/execution/plan/SingleFileTreeElementMatcher.java

```diff
@@ -131,7 +131,13 @@ public int getMode() {
 
         @Override
         public FilePermissions getPermissions() {
-            int unixNumeric = stat.getUnixMode(file);
+            File target;
+            try {
+                target = file.getCanonicalFile();
```

Comment:
❓ I just remembered that we did not use `getCanonicalFile()` in order to check path traversal issues because it was too slow.

And given what we are doing here for _all_ copied or archived files, I am afraid we end up with a problematic performance impact. Is there a way to do this only when we see a symlink?

Reply:
🤔 Would it help if we checked first if the path points to a symlink before we call `getCanonicalFile()`?

Reply:
Note: that this thread is obsolete with the latest changes.

Reply:
The initial approach with path canonicalization is 5% regression, but the latest version passes all perf tests